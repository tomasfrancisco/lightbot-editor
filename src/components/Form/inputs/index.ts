export * from "./FormLinkInput";
export * from "./FormTextInput";
export * from "./TitleInput";
export * from "./FormColorInput";
export * from "./FormBatchInput";
export * from "./FormImageInput";
export * from "./SearchInput";
export * from "./FormTextTagInput";
export * from "./FormTextAreaInput";
