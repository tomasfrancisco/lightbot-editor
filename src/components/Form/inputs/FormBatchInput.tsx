import { Icon, Input } from "antd";
import { FormComponentProps } from "antd/lib/form";
import { ValidationRule } from "antd/lib/form/Form";
import * as React from "react";
import { Button } from "src/components/Button";
import { FormItem } from "src/components/Form";
import { getUniqueNumberForSession } from "src/utils";

const { TextArea } = Input;

const BATCH_FIELD_ID = "batchField";

export type BatchInput = {
  id: number;
  value: string;
};

export type FormBatchInputProps = FormComponentProps & {
  onAddBatchFields: (newValues: BatchInput[]) => void;
  rules?: ValidationRule[];
};

type FormBatchInputState = {
  isBatchFieldOpened: boolean;
};

export class FormBatchInput extends React.Component<FormBatchInputProps, FormBatchInputState> {
  constructor(props) {
    super(props);

    this.state = {
      isBatchFieldOpened: false,
    };
  }

  public render() {
    const { isBatchFieldOpened } = this.state;

    return isBatchFieldOpened
      ? this.renderBatchInputFieldOpened()
      : this.renderBatchInputFieldClosed();
  }

  private renderBatchInputFieldClosed = () => {
    return (
      <FormItem>
        <Button onClick={this.onAddBatch} level="secondary" block={true}>
          <Icon type="plus" />
          Add Batch
        </Button>
      </FormItem>
    );
  };

  private renderBatchInputFieldOpened = () => {
    const { form, rules } = this.props;

    return (
      <div>
        <FormItem>
          {form.getFieldDecorator(BATCH_FIELD_ID, {
            rules,
            validateFirst: true,
            validateTrigger: [],
          })(<TextArea placeholder="Insert here your keywords list" />)}
        </FormItem>
        <FormItem>
          <Button onClick={this.addBatchKeywordsToList} level="secondary" block={true}>
            <Icon type="plus" />
            Add list of keywords
          </Button>
        </FormItem>
        <FormItem>
          <Button onClick={this.onCloseBatch} level="secondary" block={true}>
            <Icon type="minus" />
            Close Batch
          </Button>
        </FormItem>
      </div>
    );
  };

  private addBatchKeywordsToList = () => {
    const { form, onAddBatchFields } = this.props;
    form.validateFields(
      [BATCH_FIELD_ID],
      {
        first: true,
      },
      (errors, values) => {
        if (!errors || errors.length === 0) {
          const cleanedKeywords = new Array<BatchInput>();
          const separatedLines = values.batchField.match(/[^\r\n]+/g);
          for (const line of separatedLines) {
            cleanedKeywords.push({ id: getUniqueNumberForSession(), value: line.trim() });
          }
          form.resetFields([BATCH_FIELD_ID]);

          this.setState({
            isBatchFieldOpened: false,
          });

          onAddBatchFields(cleanedKeywords);
        }
      },
    );
  };

  private onAddBatch = () => {
    this.setState({
      isBatchFieldOpened: true,
    });
  };

  private onCloseBatch = () => {
    this.setState({
      isBatchFieldOpened: false,
    });
  };
}
