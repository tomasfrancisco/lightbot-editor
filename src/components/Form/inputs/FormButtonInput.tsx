import { Input, Select } from "antd";
import { css } from "emotion";
import _clone from "lodash.clone";
import * as React from "react";
import { FormCol, FormRow } from "~/components/Form/layout";
import { Intent } from "~/models";

const selectStyle = css`
  width: 300px;
  margin-right: 5px;
`;

export type FormButton = {
  label: string;
  intentId: string;
};

type FormButtonInputProps = {
  intents?: Intent[];
  value: FormButton;
  onChange(value: FormButton): void;
};

export class FormButtonInput extends React.Component<FormButtonInputProps> {
  public render() {
    const { value } = this.props;

    if (!value) {
      return null;
    }

    return (
      <FormRow>
        <FormCol span={12}>
          <Input
            placeholder="Button title"
            value={value.label}
            onChange={this.getOnChange("label")}
          />
        </FormCol>
        <FormCol span={12}>
          <Select
            placeholder="Jump to intent"
            defaultValue={value.intentId !== "" ? value.intentId : undefined}
            onChange={this.onChangeIntentSelector("intentId")}
            allowClear={false}
            className={selectStyle}
          >
            {this.renderSelectOptions}
          </Select>
        </FormCol>
      </FormRow>
    );
  }

  private get renderSelectOptions() {
    const { intents } = this.props;
    if (!intents) {
      return null;
    }

    return intents.map(intent => {
      return (
        <Select.Option value={intent.id} key={intent.id}>
          {intent.name}
        </Select.Option>
      );
    });
  }

  private onChangeIntentSelector = buttonProp => selectedIntentId => {
    const { onChange } = this.props;

    const propsValue = _clone(this.props.value);
    propsValue[buttonProp] = selectedIntentId;

    if (onChange) {
      onChange(propsValue);
    }
  };

  private getOnChange = buttonProp => ({ target: { value } }) => {
    const { onChange } = this.props;

    const propsValue = _clone(this.props.value);
    propsValue[buttonProp] = value;

    if (onChange) {
      onChange(propsValue);
    }
  };
}
