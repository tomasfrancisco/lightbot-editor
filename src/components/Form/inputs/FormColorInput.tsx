import { Icon, Input } from "antd";
import Form, { FormComponentProps, ValidationRule } from "antd/lib/form";
import * as React from "react";
import { SketchPicker } from "react-color";
import {
  ColorPickerBackgroundView,
  ColorPickerContainer
} from "~/components/ColorPicker/layout";
import {
  generateRgbaString,
  generateRgbString
} from "~/components/ColorPicker/utils";
import { FormItem } from "~/components/Form/components/FormItem";
import {
  closeButtonStyle,
  CloseContainer,
  ColorRow,
  ColorSelector,
  inputStyle
} from "~/components/Form/layout";
import { colorValidator } from "~/components/Form/validators";
import { ColorType } from "~/models";

export type FormColorInputProps = FormComponentProps & {
  required?: boolean;
  itemKey: string;
  title?: string;
  colorType: ColorType;
  itemGroup?: string;
  rules?: ValidationRule[];
  initialValue?: string;
  defaultValue?: string;
  placeholder?: string;
  itemKeyOpened?: string;
  onItemKeySelect(itemKey?: string): void;
};

export type FormColorInputState = {
  currentSelectedColor?: string;
};

export class FormColorInput extends React.Component<
  FormColorInputProps,
  FormColorInputState
> {
  constructor(props: FormColorInputProps) {
    super(props);

    this.state = {
      currentSelectedColor: props.initialValue || props.defaultValue
    };
  }

  public render() {
    const {
      title,
      form,
      initialValue,
      rules,
      placeholder,
      colorType,
      required
    } = this.props;
    const { currentSelectedColor } = this.state;
    const remainRules = rules ? rules : [];

    if (form) {
      return (
        <div>
          {this.renderPicker}
          <FormItem title={title}>
            <ColorRow>
              {form.getFieldDecorator(this.fieldDecoratorId, {
                initialValue,
                rules: [
                  {
                    message: "Please select some color.",
                    required: required ? true : false,
                    whitespace: true
                  },
                  colorValidator("Please insert the correct format", colorType),
                  ...remainRules
                ],
                validateFirst: true,
                validateTrigger: ["onChange", "onBlur"]
              })(
                <Input
                  className={inputStyle}
                  placeholder={placeholder}
                  onChange={this.onInputFieldChange}
                />
              )}
              <ColorSelector
                style={{ backgroundColor: currentSelectedColor }}
                onClick={this.onColorSelectorPress}
              />
            </ColorRow>
          </FormItem>
        </div>
      );
    } else {
      return null;
    }
  }

  private get renderPicker() {
    const { colorType, itemKey, itemKeyOpened } = this.props;
    const { currentSelectedColor } = this.state;
    if (itemKey === itemKeyOpened) {
      return (
        <ColorPickerBackgroundView onClick={this.onBackgroundClick}>
          <CloseContainer>
            <Icon type="close" className={closeButtonStyle} />
          </CloseContainer>
          <ColorPickerContainer onClick={this.onSketchClick}>
            <SketchPicker
              disableAlpha={colorType !== ColorType.RGBA}
              onChangeComplete={this.onChangeColor}
              color={currentSelectedColor}
            />
          </ColorPickerContainer>
        </ColorPickerBackgroundView>
      );
    } else {
      return null;
    }
  }

  private get fieldDecoratorId() {
    const { itemKey, itemGroup } = this.props;

    if (itemGroup) {
      return `${itemGroup}[${itemKey}]`;
    } else {
      return itemKey;
    }
  }

  private onBackgroundClick = () => {
    const { onItemKeySelect } = this.props;
    onItemKeySelect(undefined);
  };

  private onSketchClick = e => {
    e.stopPropagation();
  };

  private onChangeColor = color => {
    const { form } = this.props;
    let newColor = color.hex;
    const { colorType } = this.props;
    if (colorType === ColorType.RGBA) {
      newColor = generateRgbaString(color.rgb);
    } else if (colorType === ColorType.RGB) {
      newColor = generateRgbString(color.rgb);
    }

    form.setFieldsValue({ [this.fieldDecoratorId]: newColor });

    this.setState({
      currentSelectedColor: newColor
    });
  };

  private onInputFieldChange = e => {
    const { defaultValue } = this.props;

    let newValue = e.target.value;

    if (newValue === "") {
      newValue = defaultValue;
    }

    this.setState({
      currentSelectedColor: newValue
    });
  };

  private onColorSelectorPress = () => {
    const { onItemKeySelect, itemKey } = this.props;

    onItemKeySelect(itemKey);
  };
}
