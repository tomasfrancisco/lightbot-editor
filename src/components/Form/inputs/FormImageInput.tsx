import { Input, Row } from "antd";
import { FormComponentProps, ValidationRule } from "antd/lib/form";
import * as React from "react";
import Img from "react-image";
import { FormItem } from "src/components/Form/components/FormItem";
import { formInputImageStyle, ImageRow, InputImageContainer } from "src/components/Form/layout";

export type FormImageInputProps = FormComponentProps & {
  required?: boolean;
  title?: string;
  itemGroup?: string;
  itemKey: string;
  rules?: ValidationRule[];
  initialValue: string;
  placeholder?: string;
};

export type FormImageInputState = {
  src: string;
};

export class FormImageInput extends React.Component<FormImageInputProps, FormImageInputState> {
  constructor(props: FormImageInputProps) {
    super(props);

    this.state = {
      src: props.initialValue,
    };
  }

  public render() {
    const { form, initialValue, rules, placeholder, title, required } = this.props;
    const { src } = this.state;
    const remainRules = rules ? rules : [];

    if (form) {
      return (
        <FormItem title={title}>
          <Row>
            {form.getFieldDecorator(this.fieldDecoratorId, {
              initialValue,
              rules: [
                {
                  message: "Please insert some value",
                  required: required ? true : false,
                  whitespace: true,
                },
                ...remainRules,
              ],
              validateFirst: true,
              validateTrigger: ["onChange", "onBlur"],
            })(<Input placeholder={placeholder} onChange={this.onInputChange} />)}
          </Row>
          <ImageRow>
            <InputImageContainer>
              <Img className={formInputImageStyle} src={src} />
            </InputImageContainer>
          </ImageRow>
        </FormItem>
      );
    } else {
      return null;
    }
  }

  private get fieldDecoratorId() {
    const { itemKey, itemGroup } = this.props;

    if (itemGroup) {
      return `${itemGroup}[${itemKey}]`;
    } else {
      return itemKey;
    }
  }

  private onInputChange = e => {
    this.setState({ src: e.target.value });
  };
}
