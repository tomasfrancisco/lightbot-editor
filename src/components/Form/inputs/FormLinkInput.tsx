import { Input } from "antd";
import { FormComponentProps } from "antd/lib/form";
import _clone from "lodash.clone";
import _get from "lodash.get";
import * as React from "react";
import { FormCol, FormRow } from "~/components/Form/layout";

export type FormLink = {
  label: string;
  link: string;
};

type FormLinkInputProps = {
  value?: FormLink;
  onChange?(value: FormLink): void;
};

class FormLinkInput extends React.Component<FormLinkInputProps> {
  public render() {
    const { value } = this.props;

    return (
      <FormRow>
        <FormCol span={12}>
          <Input
            placeholder="Label"
            value={_get(value, "label", "")}
            onChange={this.getOnChange("label")}
          />
        </FormCol>
        <FormCol span={12}>
          <Input
            placeholder="https://example.com"
            value={_get(value, "link", "")}
            onChange={this.getOnChange("link")}
          />
        </FormCol>
      </FormRow>
    );
  }

  private getOnChange = linkProp => ({ target: { value } }) => {
    const { onChange } = this.props;
    if (!onChange) {
      return;
    }

    const propsValue = _clone(this.props.value) || {
      label: "",
      link: ""
    };

    propsValue[linkProp] = value;

    if (onChange) {
      onChange(propsValue);
    }
  };
}

export type RenderFormLinkInputProps = FormComponentProps & {
  itemKey: string;
  value?: any;
  errorMessage?: string;
  type?: string;
  getValueFromEvent?(e): any;
  getValueProps?(value): any;
  validator?(
    rule: any,
    value: any,
    callback: any,
    source?: any,
    options?: any
  ): any;
};

export const renderFormLinkInput = ({
  itemKey,
  form,
  value,
  errorMessage,
  validator,
  getValueFromEvent,
  getValueProps,
  type
}: RenderFormLinkInputProps) =>
  form &&
  form.getFieldDecorator(itemKey, {
    getValueFromEvent,
    // @ts-ignore
    getValueProps,
    initialValue: value,
    rules: [
      {
        message: errorMessage || "Provide a label and a valid URL.",
        required: true,
        type,
        validator
      }
    ],
    validateTrigger: ["onChange", "onBlur"]
  })(<FormLinkInput />);
