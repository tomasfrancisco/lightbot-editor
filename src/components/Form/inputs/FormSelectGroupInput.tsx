import { Button, Icon, Select } from "antd";
import { FormComponentProps } from "antd/lib/form";
import _clone from "lodash.clone";
import _find from "lodash.find";
import _get from "lodash.get";
import * as React from "react";
import { FormCol, FormRow } from "~/components/Form/layout";

type FormSelectGroupInputProps = {
  value?: string[];
  searchResults: Array<{ key: string; value: string }>;
  onChange?(value: string[]): void;
  onFetch?(): void;
};

type FormSelectGroupInputState = {
  hasThird: boolean;
};

export enum FormSelectGroupEnum {
  FIRST = 0,
  SECOND = 1,
  THIRD = 2
}

class FormSelectGroupInput extends React.Component<
  FormSelectGroupInputProps,
  FormSelectGroupInputState
> {
  constructor(props) {
    super(props);

    this.state = {
      hasThird: _get(props, "value", []).length === 3
    };
  }

  public render() {
    const { value } = this.props;
    const { hasThird } = this.state;

    if (!value) {
      return null;
    }

    return (
      <FormRow>
        <FormCol span={8}>
          <Select
            onChange={this.getOnChangeHandler(FormSelectGroupEnum.FIRST)}
            showSearch={true}
            placeholder="Keyword"
            optionFilterProp="children"
            value={value[FormSelectGroupEnum.FIRST]}
            allowClear={true}
            filterOption={true}
            onFocus={this.getOnFocusHandler(FormSelectGroupEnum.FIRST)}
          >
            {this.renderSearchResults()}
          </Select>
        </FormCol>
        <FormCol span={8}>
          <Select
            onChange={this.getOnChangeHandler(FormSelectGroupEnum.SECOND)}
            showSearch={true}
            placeholder="Keyword"
            optionFilterProp="children"
            value={value[FormSelectGroupEnum.SECOND]}
            allowClear={true}
            filterOption={true}
            onFocus={this.getOnFocusHandler(FormSelectGroupEnum.SECOND)}
          >
            {this.renderSearchResults()}
          </Select>
        </FormCol>
        <FormCol span={8}>
          {(hasThird && (
            <Select
              onChange={this.getOnChangeHandler(FormSelectGroupEnum.THIRD)}
              showSearch={true}
              placeholder="Keyword"
              optionFilterProp="children"
              value={value[FormSelectGroupEnum.THIRD]}
              allowClear={true}
              filterOption={true}
              onFocus={this.getOnFocusHandler(FormSelectGroupEnum.THIRD)}
            >
              {this.renderSearchResults()}
            </Select>
          )) || (
            <Button
              type="dashed"
              style={{ width: "100%" }}
              onClick={this.onAddThirdClick}
            >
              <Icon type="plus" />
            </Button>
          )}
        </FormCol>
      </FormRow>
    );
  }

  private renderSearchResults() {
    const { searchResults } = this.props;

    return searchResults.map(searchResult => (
      <Select.Option key={searchResult.key} value={searchResult.key}>
        {searchResult.value}
      </Select.Option>
    ));
  }

  private onAddThirdClick = () => this.setState({ hasThird: true });

  private getOnChangeHandler = (step: FormSelectGroupEnum) => stepValue => {
    const { onChange, searchResults } = this.props;
    if (!onChange) {
      return;
    }

    let value = _clone(this.props.value);
    if (!value) {
      value = new Array(3).fill("");
    }

    const newValue = _get(
      _find(searchResults, ["key", stepValue]),
      "value",
      ""
    );
    value[step] = newValue;

    if (onChange) {
      onChange(value);
    }
  };

  private getOnFocusHandler = (step: FormSelectGroupEnum) => () => {
    const { onFetch } = this.props;
    if (onFetch) {
      onFetch();
    }
  };
}

export type RenderFormSelectGroupInputProps = FormComponentProps & {
  value: any;
  itemKey: string;
  errorMessage?: string;
  searchResults: Array<{ key: string; value: string }>;
  min: number;
  type: string;
  getValueFromEvent?(e): any;
  getValueProps?(value): any;
  validator?(
    rule: any,
    value: any,
    callback: any,
    source?: any,
    options?: any
  ): any;
  onChange?(value: string[]): void;
  onFetch?(): void;
};

/**
 * This component should be called as a function and not as a JSX component
 * Otherwise, the form is not going to recognize this field and not throw errors
 */
export const renderFormSelectGroupInput = ({
  form,
  itemKey,
  errorMessage,
  value,
  searchResults,
  onFetch,
  getValueFromEvent,
  getValueProps,
  validator,
  min,
  type
}: RenderFormSelectGroupInputProps) =>
  form &&
  form.getFieldDecorator(itemKey, {
    getValueFromEvent,
    // @ts-ignore
    getValueProps,
    initialValue: value,
    rules: [
      {
        message: errorMessage,
        min,
        required: true,
        type,
        validator
      }
    ],
    trigger: "onChange",
    validateTrigger: ["onChange", "onBlur"]
  })(<FormSelectGroupInput onFetch={onFetch} searchResults={searchResults} />);
