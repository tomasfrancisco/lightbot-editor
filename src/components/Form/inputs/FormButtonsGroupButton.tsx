import { Icon } from "antd";
import _clone from "lodash.clone";
import * as React from "react";
import styled from "react-emotion";
import { Button } from "~/components/Button";
import { FormItem } from "~/components/Form/components";
import { FormButton, FormButtonInput } from "~/components/Form/inputs/FormButtonInput";
import { FormRow } from "~/components/Form/layout";
import theme from "~/config/theme.js";
import { Intent } from "~/models";

const Wrapper = styled("div")`
  border: 1px dashed ${theme["default-border-color"]};
  border-radius: 5px;
  padding: 5px;
`;

type FormButtonsGroupButtonsProps = {
  intents?: Intent[];
  value?: FormButton[];
  onChange?(value: FormButton[]): void;
};

export class FormButtonsGroupButtons extends React.Component<FormButtonsGroupButtonsProps> {
  public render() {
    const { intents, value } = this.props;

    if (!value) {
      return null;
    }

    return (
      <Wrapper>
        {value.map((button, index) => (
          <FormItem key={index} onRemove={this.getOnRemoveHandler(index)}>
            <FormButtonInput
              intents={intents}
              key={index}
              onChange={this.getOnChangeHandler(index)}
              value={button}
            />
          </FormItem>
        ))}
        <FormRow>
          <Button onClick={this.onAdd} level="secondary" block={true}>
            <Icon type="right-square-o" />
            Add intent jump
          </Button>
        </FormRow>
      </Wrapper>
    );
  }

  private getOnChangeHandler = index => value => {
    const { onChange } = this.props;

    const propsValue = _clone(this.props.value) || [];

    propsValue[index] = value;

    if (onChange) {
      onChange(propsValue);
    }
  };

  private getOnRemoveHandler = index => () => {
    const { onChange } = this.props;

    const propsValue = _clone(this.props.value) || [];
    const newValue = [...propsValue.slice(0, index), ...propsValue.slice(index + 1)];

    if (onChange) {
      onChange(newValue);
    }
  };

  private onAdd = () => {
    const { onChange } = this.props;

    const propsValue = _clone(this.props.value) || [];

    propsValue.push({
      intentId: "",
      label: "",
    });

    if (onChange) {
      onChange(propsValue);
    }
  };
}
