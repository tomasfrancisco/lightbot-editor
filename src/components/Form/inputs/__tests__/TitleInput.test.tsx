import { Form as AntdForm } from "antd";
import { shallow } from "enzyme";
import * as React from "react";
import { TitleInput } from "~/components/Form/inputs";
import { ColorType } from "~/models";

describe("components/Form/inputs/TitleInput", () => {
  test("renders", () => {
    const onItemkeySelect = () => {
      /**/
    };
    const Wrapper = AntdForm.create()(TitleInput);
    const tree = shallow(
      <Wrapper placeholder="placeholder" itemKey="itemkey" value="value" />
    );
    expect(tree).toMatchSnapshot();
  });
});
