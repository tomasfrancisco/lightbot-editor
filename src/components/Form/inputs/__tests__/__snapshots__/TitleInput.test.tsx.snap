// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form/inputs/TitleInput renders 1`] = `
<Component
  form={
    Object {
      "getFieldDecorator": [Function],
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldProps": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "isSubmitting": [Function],
      "resetFields": [Function],
      "setFields": [Function],
      "setFieldsInitialValue": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
      "validateFieldsAndScroll": [Function],
    }
  }
  itemKey="itemkey"
  placeholder="placeholder"
  value="value"
/>
`;
